crates/pallet-domains/src/staking.rs

```diff
@@ -387,119 +599,357 @@ pub(crate) fn do_deregister_operator<T: Config>(
+        ensure!(
+            unlock_at_confirmed_domain_block_number <= latest_confirmed_block_number,
+            Error::UnlockPeriodNotComplete
+        );
+
+        let staked_hold_id = T::HoldIdentifier::staking_staked(operator_id);
+        let locked_amount = T::Currency::balance_on_hold(&staked_hold_id, &nominator_id);
+        let amount_to_unlock: BalanceOf<T> = {
+            let epoch_share_price =
+                OperatorEpochSharePrice::<T>::get(operator_id, allowed_since_domain_epoch)
+                    .ok_or(Error::MissingOperatorEpochSharePrice)?;
+
+            // if the share price is one, just convert shares to ssc
+            let amount_to_unlock = if epoch_share_price.is_one() {
+                shares.into()
+            } else {
+                epoch_share_price.shares_to_stake::<T>(shares)
+            };
+
+            // if the amount to release is more than currently locked,
+            // mint the diff and release the rest
+            if let Some(amount_to_mint) = amount_to_unlock.checked_sub(&locked_amount) {
+                // mint any gains
+                mint_funds::<T>(&nominator_id, amount_to_mint)?;
+                locked_amount
+            } else {
+                amount_to_unlock
             }
+        };
+
+        T::Currency::release(
+            &staked_hold_id,
+            &nominator_id,
+            amount_to_unlock,
+            Precision::Exact,
+        )
+        .map_err(|_| Error::RemoveLock)?;
+
+        Ok(amount_to_unlock)
+    })
+}
+
+/// Converts shares to ssc based on the OPerator epoch share price
+/// if the share price is not available, then return the shares as is
+fn convert_shares_to_ssc<T: Config>(
+    operator_id: OperatorId,
+    domain_epoch: DomainEpoch,
+    shares: T::Share,
+) -> (BalanceOf<T>, T::Share) {
+    match OperatorEpochSharePrice::<T>::get(operator_id, domain_epoch) {
+        None => (Zero::zero(), shares),
+        Some(epoch_share_price) => {
+            // if the share price is one, just convert shares to ssc
+            (
+                if epoch_share_price.is_one() {
+                    shares.into()
+                } else {
+                    epoch_share_price.shares_to_stake::<T>(shares)
+                },
+                Zero::zero(),
+            )
         }
+    }
+}
+// If there are any withdrawals for previous epoch, then calculate the SSC for those shares
+// using the share price at which the withdraw was initiated.
+// If there are any withdrawals during the current epoch, it share price is not calculated yet
+// so just return the shares as is
+pub(crate) fn calculate_withdraw_share_ssc<T: Config>(
+    operator_id: OperatorId,
+    nominator_id: NominatorId<T>,
+) -> (BalanceOf<T>, T::Share) {
+    let withdrawals = Withdrawals::<T>::take(operator_id, nominator_id.clone()).unwrap_or_default();
+    withdrawals.into_iter().fold(
```

Comment:
Since we still have limit to the number of nominators per operator, adding another limit to the number of withdrawals per nominator can make `calculate_withdraw_share_ssc` bounded.

But for completely removing both the nominator and withdrawals limit, we need more thinking, some initial ideas:
- For the unlocked operator, it may be okay to treat the nominator's deposit as an unlocked withdrawal, and require the nominator to manually call `unlock_funds` to unlock all its funds, instead of automatically handling that inside `unlock_operator` for them.
- For slashing, since the staking fund is still inside the nominator's account it seems there is no way to move all the staking fund to the treasury without iterating each of the nominators. So we may have to introduce a sub-account per operator to handle the staking fund (like to storage fund), and simply move all this account's fund to the treasury when slashing.

Reply:
This problem is mainly due to a lazy nominator and compute should be paid by that lazy nominator itself. There is an approch that I have in mind I would like to propose but that is not a blocker for this PR since that needs more feedback from the team and a spec change as well.

I would like to get this PR in so that we start testing and this issue is not something of a big conern for me at this stage but we should be able to handle this before mainnet

Reply:
>adding another limit to the number of withdrawals per nominator can make calculate_withdraw_share_ssc bounded.

How about implementing this first the change should be small otherwise `calculate_withdraw_share_ssc` is unbounded which will be a blocker for benchmarking I believe.

Reply:
No, I would rather finish the withdrawal issue and then move to benchmarking. Both of which are not a big concern for the next testnet but rather big concern for Mainnet

Reply:
> adding another limit to the number of withdrawals per nominator can make calculate_withdraw_share_ssc bounded.

Also this is a big no for me personally where we tend to introduce these limits only to come back and change it with more invasive changes. I would rather approach these with constant time changes without any intermediate limits such as these only to change it later. Also another goal for this pr to remove existing limits and introducing new limits is just defeating the purpose :)

Reply:
Aha, since you are the one working on the benchmarking and the followup if it is not a blocker for you I have no reason to hold on.